packages/menu-surface/index.tsx

```diff
@@ -0,0 +1,357 @@
+// The MIT License
+//
+// Copyright (c) 2018 Google, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+import * as React from 'react';
+import * as classnames from 'classnames';
+import {MDCMenuSurfaceFoundation, Corner} from '@material/menu-surface';
+
+export interface MenuSurfaceProps {
+  className: string,
+  anchorElement: HTMLElement,
+  anchorCorner: number,
+  anchorMargin: object,
+  styles: object,
+  coordinates: {
+    x?: number,
+    y?: number
+  },
+  onClose: () => void,
+  onOpen: () => void,
+  onKeyDown: (event: React.KeyboardEvent) => void,
+  quickOpen: boolean,
+  open: boolean,
+  fixed: boolean
+};
+
+type MenuSurfaceState = {
+  transformOrigin: string,
+  maxHeight?: number,
+  styleLeft?: number,
+  styleRight?: number,
+  styleTop?: number,
+  styleBottom?: number,
+  classList: Set<string>
+};
+
+class MenuSurface extends React.Component<MenuSurfaceProps & React.HTMLProps<HTMLDivElement>, MenuSurfaceState> {
+  menuSurfaceElement_: React.RefObject<HTMLDivElement> = React.createRef();
+  previousFocus_: HTMLElement | null = null;
+  foundation_: MDCMenuSurfaceFoundation;
+  handleWindowClick_: EventListener;
+  registerWindowClickListener_: () => void;
+  deregisterWindowClickListener_: () => void;
+  firstFocusableElement_: HTMLElement | null;
+  lastFocusableElement_: HTMLElement | null;
+
+  state = {
+    transformOrigin: '',
+    maxHeight: undefined,
+    styleLeft: undefined,
+    styleRight: undefined,
+    styleTop: undefined,
+    styleBottom: undefined,
+    classList: new Set(),
+  } as MenuSurfaceState;
```

Comment:
I don't *think* this cast is necessary, I believe the `React.Component` typing should already type this field as `MenuSurfaceState` and check against the value you're setting it to

Reply:
What about 
``` state: MenuSurfaceState = {
    transformOrigin: '',
    maxHeight: undefined,
    styleLeft: undefined,
    styleRight: undefined,
    styleTop: undefined,
    styleBottom: undefined,
    classList: new Set(),
  };
```

Reply:
Should work fine, but still redundant to the underlying types I think (`React.Component` declares the state field with the right type already, provided your generic parameters are correct). Not a problem to re-state it explicitly though, if you think it makes the code clearer.

Reply:
Ah I changed it to this as Amr said its better to be more strict. But from the lens of React it is redundant. I think I'm going to keep it since its in all of the components right now. If we decide to remove them in the future, I can do it in one PR